src/Render/TagHooks.php

```diff
@@ -64,7 +80,16 @@ public function renderForm( $input, array $args, Parser $parser, PPFrame $frame
 		global $wgUser, $wgEmailConfirmToEdit, $IP, $wgScript;
 		$ret = '';
 		//$parser->getOutput()->addModuleStyles( 'ext.wsForm.general.styles' );
-
+		$renderonlyapprovedforms = Config::getConfigVariable( 'renderonlyapprovedforms' );
+		$renderi18nErrorInsteadofImageForApprovedForms = Config::getConfigVariable( 'renderi18nErrorInsteadofImageForApprovedForms' );
+		if ( $renderonlyapprovedforms === false ) {
+			$this->officialForm = true;
+		}
+		if ( $this->officialForm === null ) {
```

Comment:
The users in the allowedGroups settings are responsible for any form on a page, not just one. If they approve a page to have valid forms, then all forms on one page will be valid.

Reply:
But this is for rendering forms. If an approved form is rendered, then `$this->officialForm` is set to true and all forms afterwards are rendered unconditionally (so an unapproved form can be added to a page after an approved form, and both will be rendered).

Reply:
No, when an unauthorized users adds a form to an already approved page, all forms on the page are invalid, since a non approved user has edited a form, regardless of how many forms there are on a page. Everytime it runs through the form parser function, official form unset.

Reply:
I get that all forms are invalidated when someone edits a page, but this is when people display a page. Suppose I have forms "A" (approved), "B" (unapproved) and "C" (unapproved). If I then render them on a page "D", like so, all forms will be rendered:

```
{{A}} <!-- $this->officialForm is set to "true" -->
{{B}} <!-- $this->officialForm is set, so it remains "true" and the form is rendered -->
{{C}} <!-- Same as above -->
```

It doesn't seem like `$this->officialForm` gets reset for each invocation of the parser tag.